### PR TITLE
feat: add defragment command

### PR DESCRIPTION
**Pull request by BorysTheDev**

### CODE REVIEW

File: src/server/engine_shard_set.cc [hunk 1]
```diff
@@ -265,7 +265,13 @@ bool EngineShard::DefragTaskState::CheckRequired() {
   return false;
 }
 
-bool EngineShard::DoDefrag() {
+void EngineShard::ForceDefrag() {
+  DefragTaskState defrag_state_;
+  while (DoDefrag(defrag_state_))
```
Comment (romange) [inline src/server/engine_shard_set.cc:270]: this won't do. it blocks the entire server. 
please note that you only need it run once (to start it) and then CheckRequired will continue running it in background due to `cursor > 0`

Comment (BorysTheDev) [reply src/server/engine_shard_set.cc:270]: thx, it looks I misunderstand the task a little bit

File: src/server/engine_shard_set.cc [hunk 2]
```diff
@@ -251,20 +253,36 @@ bool EngineShard::DefragTaskState::CheckRequired() {
...
   }
 
-  const std::size_t threshold_mem = memory_per_shard * GetFlag(FLAGS_mem_defrag_threshold);
-  const double waste_threshold = GetFlag(FLAGS_mem_defrag_waste_threshold);
+  const std::size_t global_threshold = max_memory_limit * GetFlag(FLAGS_mem_defrag_threshold);
+  if (global_threshold > rss_mem_current.load(memory_order_relaxed)) {
+    return false;
+  }
+
+  const auto now = std::chrono::steady_clock::now();
```
Comment (romange) [inline src/server/engine_shard_set.cc:261]: no need to use `steady_clock` which is relative expensive.
use `time(nullptr)`  which gives you second precision and it's enough

File: src/server/engine_shard_set.h [hunk 3]
```diff
@@ -183,11 +183,15 @@ class EngineShard {
...
   TxQueueInfo AnalyzeTxQueue() const;
 
+  void ForceDefrag();
+
  private:
   struct DefragTaskState {
     size_t dbid = 0u;
     uint64_t cursor = 0u;
-    bool underutilized_found = false;
+    std::chrono::time_point<std::chrono::steady_clock> prev_check =
```
Comment (romange) [inline src/server/engine_shard_set.h:192]: pls use `time_t` and you can set to 0, no need to call a function here.

File: src/server/engine_shard_set.h [hunk 4]
```diff
@@ -183,11 +183,14 @@ class EngineShard {
...
   TxQueueInfo AnalyzeTxQueue() const;
 
+  void ForceDefrag();
+
  private:
   struct DefragTaskState {
     size_t dbid = 0u;
     uint64_t cursor = 0u;
-    bool underutilized_found = false;
+    time_t prev_check = 0;
```
Comment (romange) [inline src/server/engine_shard_set.h:192]: nit: `last_check_time`

**Review by romange — APPROVED** [review]

**Pull request merged**